### PR TITLE
feat: backend rubric reuse changes

### PR DESCRIPTION
**Pull request by jansenk**
**TL;DR -** Backend changes for rubric reuse

JIRA: [EDUCATOR-5752](https://openedx.atlassian.net/browse/EDUCATOR-5752)

**What changed?**

- Added a new mixin, RubricReuseMixin, that provides rubric-reuse-related functionality
- `get_other_course_ora_blocks` returns the list of all other ORA XBlocks in the course
-  added `get_other_ora_blocks_for_rubric_editor_context`, which internally calls `get_other_course_ora_blocks` and returns a list of dicts with block Display Name and Location
- `get_rubric` returns the rubric information for an ORA XBlock given a block locator
- modify `StudioMixin.get_editor_context` to call `get_other_ora_blocks_for_rubric_editor_context` if the feature flag is enabled, and return two new keys in the context dict: 
               - `'rubric_reuse_enabled'`: is the feature enabled?
               - `'other_course_openassessments'`: the return value of `get_other_ora_blocks_for_rubric_editor_context`, unless the flag is disabled, then an empty list.
- Added opaque-keys as a requirement

**Developer Checklist**

- [x] Reviewed the [release process](https://github.com/edx/edx-ora2/blob/master/.github/release_process.md)
- [x] Translations and JS/SASS compiled
- [ ] Bumped version number in [setup.py](https://github.com/edx/edx-ora2/blob/a62e81a9b0d89223476967ec3c27f3557a850735/setup.py#L39) and [package.json](https://github.com/edx/edx-ora2/blob/a62e81a9b0d89223476967ec3c27f3557a850735/package.json#L3)

**Testing Instructions**

-I'm actually incredibly unsure!! I manually tested by adding the following lines to `studio_mixin.py:149` ( above `return { 'prompts':`)

```
        other_ora_blocks = self.get_other_course_ora_blocks()
        for other_ora_block in other_ora_blocks:
            logger.info(block.location)
            logger.info(self._get_rubric(other_ora_block.location))
```

Then, when you go to studio and edit an ORA, when you click the "edit" button you should see all other ORAs and rubrics in the logs

**Reviewer Checklist**

Collectively, these should be completed by reviewers of this PR:

- [x] I've done a visual code review
- [x] I've tested the new functionality

FYI: @edx/masters-devs-gta

### CODE REVIEW

File: openassessment/xblock/test/test_rubric_reuse_mixin.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
+""" Unit tests for rubric_reuse_mixin.py """
```
Comment (leangseu-edx) [inline openassessment/xblock/test/test_rubric_reuse_mixin.py:1]: This is a nit from other files's naming convention, I think this should name `test_rubric_reuse.py`

File: openassessment/xblock/test/test_rubric_reuse_mixin.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        self.assertFalse(resp['success'])
+        self.assertEqual(resp['msg'], 'Invalid block id.')
+
+    def test_get_rubric__not_found(self):
+        """ Test for get_rubric error behavior when the requested block can't be found """
+        with self.mock_get_ora_block(side_effect=TargetORABlockNotFoundException):
+            resp = self._request_get_rubric({'target_rubric_block_id': self.OTHER_BLOCK_LOCATION})
+        self.assertDictEqual(
+            resp,
+            self.block._get_item_error_response(self.other_block_location)
```
Comment (leangseu-edx) [inline openassessment/xblock/test/test_rubric_reuse_mixin.py:152]: `_get_item_error_response` was not mock or implemented. Was it support to be inherit from somewhere else?

**Review by leangseu-edx — CHANGES_REQUESTED** [review]

File: openassessment/xblock/rubric_reuse_mixin.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            }
+        return {
+            'success': True,
+            'rubric': {**rubric}
+        }
+
+    def _get_rubric(self, target_ora_block_locator):
+        target_block = RubricReuseMixin._get_ora_block(target_ora_block_locator)
+        if target_block.category != 'openassessment':
+            raise TargetBlockNotORAException
```
Comment (nsprenkle) [inline openassessment/xblock/rubric_reuse_mixin.py:59]: That's a _very_ specific exception 😅

File: openassessment/xblock/rubric_reuse_mixin.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+class RubricReuseMixin:
+
+    def _get_course_ora_blocks(self):
+        """
+        Get all ORA blocks from the course.
+        Both draft and published versions are queried, with preference to draft versions
+        """
+        from xmodule.modulestore.django import modulestore
+
+        return modulestore().get_items(
```
Comment (nsprenkle) [inline openassessment/xblock/rubric_reuse_mixin.py:16]: Any sense of having to bake in more course/role isolation? e.g. could non-privelaged users or admins from other courses abuse this?

I'm thinking no, but want to make sure.

File: openassessment/xblock/test/test_rubric_reuse_mixin.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+        resp = self._request_get_rubric({'target_rubric_block_id': 'invalidKEEEEY'})
+        self.assertFalse(resp['success'])
+        self.assertEqual(resp['msg'], 'Invalid block id.')
+
+    def test_get_rubric__not_found(self):
+        """ Test for get_rubric error behavior when the requested block can't be found """
+        with self.mock_get_ora_block(side_effect=TargetORABlockNotFoundException):
+            resp = self._request_get_rubric({'target_rubric_block_id': self.OTHER_BLOCK_LOCATION})
+        self.assertFalse(resp['success'])
+        self.assertRegex(
```
Comment (nsprenkle) [inline openassessment/xblock/test/test_rubric_reuse_mixin.py:144]: I had no idea we could do this 😮

File: requirements/base.txt [hunk 6]
```diff
@@ -2,8 +2,12 @@
 # This file is autogenerated by pip-compile
 # To update, run:
 #
-#    make upgrade
+#    pip-compile --output-file=requirements/base.txt requirements/base.in
 #
+--index-url http://edx.devstack.devpi:3141/root/pypi/+simple/
+--extra-index-url https://pypi.python.org/simple
+--trusted-host edx.devstack.devpi
+
```
Comment (nsprenkle) [inline requirements/base.txt:10]: These should probably get stripped just to minimize the change set

File: requirements/pip-tools.txt [hunk 7]
```diff
@@ -2,8 +2,12 @@
 # This file is autogenerated by pip-compile
 # To update, run:
 #
-#    make upgrade
+#    pip-compile --output-file=requirements/pip-tools.txt requirements/pip-tools.in
 #
+--index-url http://edx.devstack.devpi:3141/root/pypi/+simple/
+--extra-index-url https://pypi.python.org/simple
+--trusted-host edx.devstack.devpi
+
```
Comment (nsprenkle) [inline requirements/pip-tools.txt:10]: Similar notes, this didn't change meaningfully so let's not change it.

**Review by nsprenkle — APPROVED** [review]
Looks good and I've tested the 2 main functions locally (with your test instructions). See note about requirement file cleanup.

File: openassessment/xblock/rubric_reuse_mixin.py (hunk 3 above)
Comment (mattcarter) [reply openassessment/xblock/rubric_reuse_mixin.py:59]: @nsprenkle concern beyond the humor?

**Comment by jansenk** [comment]
Oh hey, that fixed it .... somehow? TIme to actually get to PR comments!

**Review by leangseu-edx — APPROVED** [review]

**Comment by jansenk** [comment]
After speaking with @ormsbee I will be switching this code over to use `xblock.runtime.modulestore` rather than importing modulestore directly

File: openassessment/xblock/test/test_studio.py [hunk 8]
```diff
@@ -5,12 +5,24 @@
...
+from unittest.mock import MagicMock, patch, Mock, PropertyMock
 
 from ddt import ddt, file_data
 import pytz
 
 from .base import XBlockHandlerTestCase, scenario
+from openassessment.xblock.config_mixin import ConfigMixin
+
+
+def disable_rubric_reuse(func):
```
Comment (jansenk) [inline openassessment/xblock/test/test_studio.py:17]: Gotta say I think this is pretty nifty

**Review by nsprenkle — APPROVED** [review]
Have done visual review and tested locally :shipit:

File: openassessment/xblock/rubric_reuse_mixin.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        Get all ORA blocks from the course.
+        Both draft and published versions are queried, with preference to draft versions
+        """
+        if hasattr(self.runtime, 'modulestore'):
+            return self.runtime.modulestore.get_items(
+                self.location.course_key,
+                qualifiers={'category': 'openassessment'},
+            )
+        else:
+            logger.info("_get_course_ora_blocks was called for {block_id} but no modulestore was available".format(
```
[the remark this answers is not in the document]
Comment (jansenk) [reply openassessment/xblock/rubric_reuse_mixin.py:24]: There are other places where we kinda expect this and have non-error behavior? 

```python
     def is_released(self, step=None):
        """
         ....
        """
        # By default, assume that we're published, in case the runtime doesn't support publish date.
        if hasattr(self.runtime, 'modulestore'):
            is_published = self.runtime.modulestore.has_published_version(self)
        else:
            is_published = True
```

File: openassessment/xblock/studio_mixin.py [hunk 10]
```diff
@@ -176,7 +181,9 @@ def editor_context(self):
             'is_released': self.is_released(),
             'teamsets': self.get_teamsets(course_id),
             'selected_teamset_id': self.selected_teamset_id,
-            'show_rubric_during_response': self.show_rubric_during_response
+            'show_rubric_during_response': self.show_rubric_during_response,
+            'rubric_reuse_enabled': self.is_rubric_reuse_enabled,
+            'other_course_openassessments': other_course_openassessments,
```
[the remark this answers is not in the document]
Comment (jansenk) [reply openassessment/xblock/studio_mixin.py:186]: good point. i'm just gonna take your name
